framework/deproxy_client.py

```diff
@@ -51,31 +56,55 @@ def handle_read(self):
             return
         tf_cfg.dbg(4, '\tDeproxy: Client: Receive response.')
         tf_cfg.dbg(5, self.response_buffer)
-        try:
-            response = deproxy.Response(self.response_buffer,
-                                method=self.request.method)
-            self.response_buffer = self.response_buffer[len(response.msg):]
-        except deproxy.IncompliteMessage:
-            return
-        except deproxy.ParseError:
-            tf_cfg.dbg(4, ('Deproxy: Client: Can\'t parse message\n'
-                           '<<<<<\n%s>>>>>'
-                           % self.response_buffer))
-            raise
-        if len(self.response_buffer) > 0:
-            # TODO: take care about pipelined case
-            raise deproxy.ParseError('Garbage after response'
-                                     ' end:\n```\n%s\n```\n' % \
-                                     self.response_buffer)
-        self.recieve_response(response)
-        self.response_buffer = ''
+        while len(self.response_buffer) > 0:
+            try:                
+                method = self.requests[self.nrresp]
+                response = deproxy.Response(self.response_buffer,
+                                    method=method)
+                self.response_buffer = \
+                            self.response_buffer[response.original_length:]
+            except deproxy.IncompliteMessage:
+                return
+            except deproxy.ParseError:
+                tf_cfg.dbg(4, ('Deproxy: Client: Can\'t parse message\n'
+                               '<<<<<\n%s>>>>>'
+                            % self.response_buffer))
+                raise
+            self.recieve_response(response)
+            self.nrresp += 1
+
+            if self.nrreq == self.nrresp and len(self.response_buffer) > 0:
```

Comment:
Imagine that `N = self.nrresp = self.nrreq` responses was received and `len(self.response_buffer) == 0`. (The condition is false) So we exit the function without errors. After that a new _part_ of response is received. It'll be caught in `except deproxy.IncompliteMessage` case. No garbage warnng will happen in this case. But garbage was received.

Imagine other situation. `N = self.nrresp = self.nrreq` responses was received and `len(self.response_buffer) == 0`. The condition is false, and no issue happen. But after that a new correct response is received. The function doesn't threat the case as issue.

Reply:
Fixed